ruby/ql/lib/codeql/ruby/ast/internal/Variable.qll

```diff
@@ -366,7 +366,23 @@ private module Cached {
 
   cached
   predicate isCapturedAccess(LocalVariableAccess access) {
-    access.getVariable().getDeclaringScope() != access.getCfgScope()
+    exists(Scope scope1, Scope scope2 |
+      scope1 = access.getVariable().getDeclaringScope() and
+      scope2 = access.getCfgScope() and
+      scope1 != scope2
+    |
+      if access instanceof SelfVariableAccess
+      then
+        // ```
+        // class C
+        //   def self.m // not a captured access
+        //   end
+        // end
+        // ```
+        not scope2 instanceof Toplevel or
+        not access = any(SingletonMethod m).getObject()
+      else any()
+    )
```

Comment:
I don't think it was introduce by this change, but I've just noticed that reads of variables captured by a lambda get interpreted as methods calls with an implicit `self`, which seems wrong:
```rb
x = 1
y = -> { x } # x is interpreted as self.x()
```

Reply:
Let's fix this follow-up.